File: python/src/semantic_retrieval/ingestion/data_sources/fs/file_system.py
import hashlib
import mimetypes
import os
import uuid
from typing import Any, Callable, List, Optional

from langchain.document_loaders import (
    CSVLoader,
    Docx2txtLoader,
    PyPDFLoader,
    TextLoader,
)
from langchain.document_loaders.base import BaseLoader
from result import Err, Ok, Result
from semantic_retrieval.common.types import CallbackEvent
from semantic_retrieval.document.document import RawDocument, RawDocumentChunk
from semantic_retrieval.ingestion.data_sources.data_source import DataSource
from semantic_retrieval.utils.callbacks import CallbackManager, Traceable
from semantic_retrieval.utils.configs.configs import remove_nones


# TODO [P1]: (suyog) I dislike this quite a bit, but following typescript for now - same with the FileSystemRawDocument implementation of RawDocument
def csv_loader_func(path: str) -> CSVLoader:
    return CSVLoader(path)


def txt_loader_func(path: str) -> TextLoader:
    return TextLoader(path)


def pdf_loader_func(path: str) -> PyPDFLoader:
    return PyPDFLoader(path)


def docx_loader_func(path: str) -> Docx2txtLoader:
    return Docx2txtLoader(path)


DEFAULT_FILE_LOADERS: dict[str, Callable[[str], BaseLoader]] = {
    ".csv": csv_loader_func,
    ".txt": txt_loader_func,
    ".json": txt_loader_func,
    ".py": txt_loader_func,
    ".md": txt_loader_func,
    ".pdf": pdf_loader_func,
    ".docx": docx_loader_func,
}


class FileSystemRawDocument(RawDocument):
    file_loaders: dict[str, Callable[[str], BaseLoader]] = DEFAULT_FILE_LOADERS

    async def get_content(self) -> Result[str, str]:
        # Get file loader w/ file_path (which is self.uri) & load_chunked_content
        _, file_extension = os.path.splitext(self.uri)

        if file_extension in self.file_loaders:
            file_loader = self.file_loaders[file_extension]
            loader = file_loader(self.uri)
            return Ok(loader.load()[0].page_content)
        else:
            return Err(f"File extension {file_extension} not supported")

    async def get_chunked_content(self) -> List[RawDocumentChunk]:
        # TODO [P1]: Implement later - not the same because lazy_load in langchain python is different
        return []


class FileSystem(DataSource, Traceable):
    name: str = "FileSystem"
    path: str
    collection_id: Optional[str] = None
    callback_manager: CallbackManager
    file_loaders: dict[str, Callable[[str], BaseLoader]] = DEFAULT_FILE_LOADERS

    def __init__(
        self,
        path: str,
        callback_manager: CallbackManager,
        collection_id: Optional[str] = None,
        file_loaders: Optional[dict[str, Callable[[str], BaseLoader]]] = None,
    ):
        self.path = path
        self.collection_id = collection_id
        self.callback_manager = callback_manager
        if file_loaders is not None:
            self.file_loaders = file_loaders

    def check_stats(self):
        return os.path.isdir(self.path), os.path.isfile(self.path)

    async def load_file(
        self, path: str, collection_id: str
    ) -> FileSystemRawDocument:
        # TODO [P1] exception handling
        file_name_with_ext = os.path.basename(path)
        file_name = os.path.splitext(file_name_with_ext)[0]
        # TODO [P1]: This should be done outside of python
        hash = hashlib.md5(open(path, "rb").read()).hexdigest()

        fsrd_args = remove_nones(
            dict(
                file_loaders=self.file_loaders,
                uri=path,
                data_source=self,
                name=file_name,
                mime_type=mimetypes.guess_type(path)[0],
                hash=hash,
                blob_id=None,
                document_id=str(uuid.uuid4()),
                collection_id=collection_id,
            )
        )

        await self.callback_manager.run_callbacks(
            CallbackEvent(
                name="file_system_document_loaded",
                data=dict(
                    path=path,
                    **fsrd_args,
                ),
            )
        )

        return FileSystemRawDocument(**fsrd_args)

    async def load_documents(
        self, filters: Optional[Any] = None, limit: Optional[int] = None
    ) -> List[RawDocument]:
        # TODO [P1]: Filters & Limit are not implemented yet

        # Iterate through directory or just load a single file & make a list, handle error conditions like can't find file or directory
        isdir, isfile = self.check_stats()
        raw_documents: List[RawDocument] = []

        if isdir:
            files = [f for f in os.listdir(self.path)]
            collection_id = (
                self.collection_id if self.collection_id else str(uuid.uuid4())
            )
            for file in files:
                subdir_path = os.path.join(self.path, file)
                if os.path.isdir(subdir_path):
                    subDir = FileSystem(
                        path=subdir_path,
                        callback_manager=self.callback_manager,
                        collection_id=collection_id,
                    )
                    raw_documents.extend(await subDir.load_documents())
                elif os.path.isfile(subdir_path):
                    raw_documents.append(
                        await self.load_file(subdir_path, collection_id)
                    )
        elif isfile:
            collection_id = (
                self.collection_id if self.collection_id else str(uuid.uuid4())
            )
            raw_documents.append(
                await self.load_file(
                    self.path,
                    collection_id,
                )
            )
        else:
            message = f"{self.path} is neither a file nor a directory."
            err = Exception(message)

            raise Exception(err)

        await self.callback_manager.run_callbacks(
            CallbackEvent(
                name="file_system_documents_loaded",
                data=dict(
                    path=self.path,
                    collection_id=self.collection_id,
                    documents=raw_documents,
                ),
            )
        )

        return raw_documents

File: typescript/src/ingestion/data-sources/fs/fileSystem.ts
import { JSONObject } from "../../../common/jsonTypes";
import { RawDocument } from "../../../document/document";
import { DataSource } from "../dataSource";
import { LangChainFileLoader } from "./langchainFileLoader";
import { TxtFileLoader } from "./txtFileLoader";
import fs from "fs/promises";
import { extname, resolve } from "node:path";
import * as mime from "mime-types";
import { v4 as uuid } from "uuid";
import { DocxFileLoader } from "./docxFileLoader";
import { PDFFileLoader } from "./pdfFileLoader";
import { CSVFileLoader } from "./csvFileLoader";
import { Md5 } from "ts-md5";

// the import from src/utils/callbacks.ts
import {
  CallbackManager,
  LoadDocumentsSuccessEvent,
  DataSourceTestConnectionErrorEvent,
  DataSourceTestConnectionSuccessEvent,
  LoadDocumentsErrorEvent,
} from "../../../utils/callbacks";

type FileLoaderMap = {
  [extension: string]: (
    path: string,
    callbackManager?: CallbackManager
  ) => LangChainFileLoader;
};

const DEFAULT_FILE_LOADERS: FileLoaderMap = {
  ".csv": (path: string, callbackManager?: CallbackManager) =>
    new CSVFileLoader(path, { callbackManager }),
  ".docx": (path: string, callbackManager?: CallbackManager) =>
    new DocxFileLoader(path, { callbackManager }),
  ".pdf": (path: string, callbackManager?: CallbackManager) =>
    new PDFFileLoader(path, { callbackManager }),
  ".txt": (path: string, callbackManager?: CallbackManager) =>
    new TxtFileLoader(path, { callbackManager }),
  // TODO: We should probably have a different loader for MD which loads sections delineated by headers, etc.
  ".md": (path: string, callbackManager?: CallbackManager) =>
    new TxtFileLoader(path, { callbackManager }),
};

export interface FileSystemConfig {
  path: string;
  collectionId?: string;
  fileLoaders?: FileLoaderMap;
  callbackManager?: CallbackManager;
}

/**
 * A data source that loads documents from a file or directory (recursive) in
 * the local file system.
 */
export class FileSystem implements DataSource, FileSystemConfig {
  name: string = "FileSystem";
  path: string;
  collectionId: string | undefined;
  callbackManager?: CallbackManager;
  fileLoaders: FileLoaderMap = {};

  constructor(config: FileSystemConfig) {
    this.path = config.path;
    this.collectionId = config.collectionId;
    this.fileLoaders = config.fileLoaders ?? DEFAULT_FILE_LOADERS;
    this.callbackManager = config.callbackManager;
  }

  private async getStats(): Promise<{
    isDirectory: () => boolean;
    isFile: () => boolean;
  }> {
    const stats = await fs.stat(this.path);
    return {
      isDirectory: () => stats.isDirectory(),
      isFile: () => stats.isFile(),
    };
  }

  private async loadFile(
    filePath: string,
    collectionId?: string
  ): Promise<RawDocument> {
    const extension = extname(filePath);
    const fileLoader = this.fileLoaders[extension];
    if (!fileLoader) {
      throw new Error(`No file loader found for extension ${extension}`);
    }

    const mimeType = mime.lookup(filePath);

    const hash = new Md5();
    const chunks = await fileLoader(
      filePath,
      this.callbackManager
    ).loadChunkedContent();
    for (const chunk of chunks) {
      hash.appendStr(chunk.content + "\n");
    }
    hash.end();

    return {
      uri: filePath,
      dataSource: this,
      name: filePath,
      mimeType: mimeType || "unknown",
      documentId: uuid(),
      collectionId,
      getChunkedContent: async () =>
        await fileLoader(filePath).loadChunkedContent(),
      getContent: async () =>
        await fileLoader(filePath)
          .loadChunkedContent()
          .then((content) => content.join("\n")),
      hash: hash.toString(),
      metadata: {},
      attributes: {},
    };
  }

  async testConnection(): Promise<number> {
    try {
      await this.getStats();
      // If stat succeeds, then the path exists.
      const out = 200;

      const event: DataSourceTestConnectionSuccessEvent = {
        name: "onDataSourceTestConnectionSuccess",
        code: out,
      };
      await this.callbackManager?.runCallbacks(event);
      return out;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (err: any) {
      const out = 404;

      const event: DataSourceTestConnectionErrorEvent = {
        name: "onDataSourceTestConnectionError",
        code: out,
        error: err,
      };
      await this.callbackManager?.runCallbacks(event);
      return out;
    }
  }

  async loadDocuments(
    _filters?: JSONObject,
    _limit?: number | undefined
  ): Promise<RawDocument[]> {
    const { isDirectory, isFile } = await this.getStats();
    const rawDocuments: RawDocument[] = [];

    // TODO: Figure out proper handling of filters and limit (likely BFS)

    if (isDirectory()) {
      const files = await fs.readdir(this.path, { withFileTypes: true });
      const collectionId = this.collectionId ?? uuid();

      const loadDirDocuments = files.map(async (file) => {
        if (file.isDirectory()) {
          const subDir = new FileSystem({
            path: resolve(this.path, file.name),
            collectionId,
          });
          return await subDir.loadDocuments();
        } else if (file.isFile()) {
          return [
            await this.loadFile(resolve(this.path, file.name), collectionId),
          ];
        }
        // If nested path is not a file or directory, just skip for now
      });

      rawDocuments.push(
        ...(await Promise.all(loadDirDocuments))
          .flat()
          .filter((doc): doc is RawDocument => doc != null)
      );
    } else if (isFile()) {
      rawDocuments.push(await this.loadFile(this.path));
    } else {
      const err = new Error(`${this.path} is neither a file nor a directory.`);

      const event: LoadDocumentsErrorEvent = {
        name: "onLoadDocumentsError",
        dataSource: this,
        error: err,
      };
      await this.callbackManager?.runCallbacks(event);
      throw err;
    }

    const event: LoadDocumentsSuccessEvent = {
      name: "onLoadDocumentsSuccess",
      dataSource: this,
      rawDocuments: rawDocuments,
    };
    await this.callbackManager?.runCallbacks(event);
    return rawDocuments;
  }
}
